### PR TITLE
Forcing column dtypes and schema check corrections

### PR DESCRIPTION
**Pull request by ilight1542**
#57 fix for column parsing and error checking

Changes made:
1) parsing of the .tsv file into a pandas dataframe now forces columns to be a specific dtype (eg string, int, or NaN if "NA" entered in .tsv) --> error will be thrown for the given line during parsing.
*Note*: This means the errors will come one at a time when doing checks (eg if you have multiple rows that have wrong datatype in the .tsv it will not tell you all of them, it will error on the first instance and abort).

2) schema checks can now interpret python NoneType as a 'null' type in json schema.
*Note*: There are some fields which have set list of accepted values defined in another json file. (eg https://spaam-community.github.io/AncientMetagenomeDir/assets/enums/library_treatment.json). Even if 'null' is accepted value, the required fields will cause error to be thrown for this entry. I recommend removing all instances of accepted type 'null' where full list of accepted values is known.

Related changes that we should considered implementing (but are outside the scope of the current PR):
1) Remove all multi-datatype column possibilities from json schema in ancientmetagenomicsdir (eg download_sizes , https://github.com/SPAAM-community/AncientMetagenomeDir/blob/master/ancientmetagenome-environmental/libraries/ancientmetagenome-environmental_libraries_schema.json) 
2) Remove accepted type 'null' from schema.json if set list of accepted values is already available.

### CODE REVIEW

**Comment by ilight1542** [comment]
import line `from xxlimited import Null` causing error in python 3.10, but not actually used anywhere.

**Comment by maxibor** [comment]
> 

Are you using github copilot ? It somestimes adds weird imports for some reason.

File: AMDirT/validate/domain/__init__.py [hunk 1]
```diff
@@ -71,16 +72,29 @@ def read_schema(self, schema: Schema) -> dict:
...
+            'integer': pd.Int64Dtype(),
+        }
+        column_dtypes = {}
+        for column_keys in schema['items']['properties']:
+            parsed_schema_dtypes=schema['items']['properties'][column_keys]['type']
+            if len(parsed_schema_dtypes) == 1:
+                column_dtypes[column_keys] = schema['items']['properties'][column_keys]['type']
+            else:
+                if parsed_schema_dtypes == ['integer','null']: column_dtypes[column_keys] = string_to_dtype_conversions['integer']
+                if parsed_schema_dtypes == ['string','null']: column_dtypes[column_keys] = string_to_dtype_conversions['string']
```
Comment (maxibor) [inline AMDirT/validate/domain/__init__.py:94]:
```suggestion
    coltype = schema['items']['properties'][column_keys]['type'][0]
    if coltype in string_to_dtype_conversions:
        column_dtypes[column_keys] = string_to_dtype_conversions[coltype]
    elif coltype == 'null':
        self.add_error(
            DFError(
                error = "Schema Error"
                source = column_dtypes[column_keys]
                column = column_keys
                row = None
                message = "Default/first type of column in schema can not be null"

        )
    else:
        column_dtypes[column_keys] = coltype
```

As null should not be the default column type in the schema, we should only care for the first element of the type list

**Review by maxibor — CHANGES_REQUESTED** [review]
Very nice @ilight1542 !
One last thing would be to add tests (you can add test datasets with Nulls in columns, and mixed data types) to see if they're properly caught

**Comment by ilight1542** [comment]
@maxibor added test for incorrect column entry (eg a numeric column has a "," in it)

The null datatype seems to mess up the parsing somehow... not sure why it causes error in column parsing(see e41e307 commit) but I think that we will always keep the NA as missing data (since pandas automatically parses this as np.NaN in the column :)

**Comment by maxibor** [comment]
I've added a few things to your PR:
- There is now a lazy reporting of the parsing (schema + dataset) errors, to get a nice table of the errors at the end.
- The tests will only be run if both the schema and dataset can be parsed respectively by python/json and pandas.
- I've removed the null type redefinition, as both in your message and in my test it does not matter.

Can you have a look at it @ilight1542 and let me know if you're fine with it ?
Otherwise, that's a ✅ for me !

**Comment by ilight1542** [comment]
Looks good to me, the changes should make any future schema changes easy to validate :)

**Review by maxibor — APPROVED** [review]

**Pull request merged**